### PR TITLE
TF text classification examples

### PR DESCRIPTION
**Pull request by gante**
# What does this PR do?

This PR updates the text classification examples to use the more modern `to_tf_dataset()` method, as opposed to a custom function. A few minor fixes were added as I navigated related files.

Example of running command: `python run_glue.py --model_name_or_path distilbert-base-cased --task_name mnli --do_train --do_eval --output_dir $HOME/test_model`

### CODE REVIEW

File: docs/source/main_classes/processors.mdx [hunk 1]
```diff
@@ -65,12 +65,7 @@ Those processors are:
 Additionally, the following method can be used to load values from a data file and convert them to a list of
 [`~data.processors.utils.InputExample`].
 
-automethod,transformers.data.processors.glue.glue_convert_examples_to_features
-
-
-### Example usage
-
-An example using these processors is given in the [run_glue.py](https://github.com/huggingface/transformers/tree/master/examples/legacy/text-classification/run_glue.py) script.
+[[autodoc]] data.processors.glue.glue_convert_examples_to_features
```
Comment (gante) [inline docs/source/main_classes/processors.mdx:68]: From my limited doc knowledge, `automethod` is from the old style. Not sure if this is the correct fix (did some pattern copy), the link to the staging docs seems broken .

File: docs/source/main_classes/processors.mdx [hunk 2]
```diff
@@ -65,12 +65,7 @@ Those processors are:
 Additionally, the following method can be used to load values from a data file and convert them to a list of
 [`~data.processors.utils.InputExample`].
 
-automethod,transformers.data.processors.glue.glue_convert_examples_to_features
-
-
-### Example usage
-
-An example using these processors is given in the [run_glue.py](https://github.com/huggingface/transformers/tree/master/examples/legacy/text-classification/run_glue.py) script.
```
Comment (gante) [inline docs/source/main_classes/processors.mdx:73]: The link was broken and I can't find any example of the GLUE processors -> deleted

File: examples/tensorflow/text-classification/run_text_classification.py [hunk 3]
```diff
@@ -399,6 +361,11 @@ def preprocess_function(examples):
         return result
 
     datasets = datasets.map(preprocess_function, batched=True, load_from_cache_file=not data_args.overwrite_cache)
+
+    if data_args.pad_to_max_length:
+        data_collator = partial(default_data_collator, return_tensors="tf")
+    else:
+        data_collator = DataCollatorWithPadding(tokenizer, return_tensors="tf")
```
Comment (gante) [inline examples/tensorflow/text-classification/run_text_classification.py:368]: an alternative to `return_tensors="tf"` is to set the collator args in `to_tf_dataset()`, but I think this way is more user-friendly

File: examples/tensorflow/text-classification/run_glue.py [hunk 4]
```diff
@@ -456,13 +417,13 @@ def compute_metrics(preds, label_ids):
...
-                non_label_column_names,
+            data = dataset.to_tf_dataset(
+                columns=[col for col in dataset.column_names if col not in set(non_label_column_names + ["label"])],
+                shuffle=shuffle,
                 batch_size=batch_size,
-                dataset_mode=dataset_mode,
+                collate_fn=data_collator,
                 drop_remainder=drop_remainder,
-                shuffle=shuffle,
+                label_cols="label" if "label" in dataset.column_names else None,
```
Comment (gante) [inline examples/tensorflow/text-classification/run_glue.py:426]: From the [docs](https://huggingface.co/docs/datasets/master/package_reference/main_classes.html#datasets.Dataset.to_tf_dataset) in `to_tf_dataset`, I would think that not setting `label_cols` would be fine for training purposes. However, training requires this field and also fails if `labels` are part of the input. Double-checking: it is the right behavior, correct?

If yes, I'm adding a PR to clear out the docs (that setting `label_cols` is required to fit). Maybe also an assertion that the label col can't be in the input cols? I'd say is a recipe for potential data leakage, having the labels in the inputs.

File: docs/source/main_classes/processors.mdx (hunk 1 above)
Comment (sgugger) [reply docs/source/main_classes/processors.mdx:68]: This is all deprecated code, but thanks for fixing the doc!

File: examples/tensorflow/text-classification/run_glue.py (hunk 4 above)
Comment (sgugger) [reply examples/tensorflow/text-classification/run_glue.py:426]: cc @Rocketknight1 for the internal of `to_tf_datatset`

**Review by sgugger — APPROVED** [review]
Thanks for working on this!

File: examples/tensorflow/text-classification/run_glue.py (hunk 4 above)
Comment (Rocketknight1) [reply examples/tensorflow/text-classification/run_glue.py:425]: This is a good question, and I should probably document it more clearly (including in the `to_tf_dataset` docs). `tf.data.Dataset` objects emit either a dict `features` or a tuple of two dicts `(features, labels)` when you iterate over them. The `columns` argument sets which columns of the input dataset will be used as keys in the `features` dict and the `label_cols` argument sets which columns of the input dataset will be used as keys in the `labels` dict. If `label_cols` is not set then the `tf.data.Dataset` will only emit a `features` dict.

The default behaviour of Keras is that it expects to get `(features, labels)` tuples as input. It does a forward pass with the `features` to get model outputs, and then passes the outputs and labels to the loss function specified in `compile()` to get a loss.

The problem with that, though, is that our models often have weird losses that are really painful for users to implement correctly - e.g. BERT has a token cross-entropy loss which should only be computed on tokens selected for masking, plus a next-sentence prediction loss. No user is ever going to correctly specify a Keras loss function for that, so we need to help them out.

The solution I used was to take advantage of the fact that our models compute a `loss` tensor during the forward pass if you pass the required labels as keys in the input dict. This is the standard way to train them with PyTorch - you just pass everything in the input dict and then do `outputs.loss.backward()`. In Keras, you can do the same thing - you just specify a "dummy" loss function that reads that loss value and outputs it directly as the loss.

So if the user `compile`s our models without a `loss` argument, we interpret that as them wanting to use the internal loss, and we quietly stick a dummy loss function on the `loss` output head of the model. If they're doing this, they have to include all the label columns as keys in the input dict (`compile()` throws a warning when you don't specify a loss to remind you to do this). If, on the other hand, the user specifies a loss when calling `compile()`, then we just use the standard Keras behaviour. In that case, it doesn't matter whether the label keys are in the input dict or not because we totally ignore the `loss` output head, but the user does have to pass a separate labels array or dict (usually by using the `label_cols` argument to `to_tf_dataset`) so that Keras can find them.

Comment (sgugger) [reply examples/tensorflow/text-classification/run_glue.py:425]: Amen to all of the above ;-)

Comment (gante) [reply examples/tensorflow/text-classification/run_glue.py:425]: @Rocketknight1 added a few more comments to attempt to summarily explain what you wrote above :)

File: src/transformers/modeling_tf_utils.py [hunk 5]
```diff
@@ -882,7 +882,15 @@ def compute_loss(self, *args, **kwargs):
 
     def train_step(self, data):
         """
-        A modification of Keras's default train_step that cleans up the printed metrics when we use a dummy loss.
+        A modification of Keras's default `train_step` that cleans up the printed metrics when we use a dummy loss. If a
```
Comment (sgugger) [inline src/transformers/modeling_tf_utils.py:885]: Docs upgrade should ideally go in a separate PR, it would be best to keep this one focused on its purpose :-)

Comment (gante) [reply src/transformers/modeling_tf_utils.py]: Moved to a separate PR 👉  https://github.com/huggingface/transformers/pull/15755

**Review by Rocketknight1 — APPROVED** [review]
